README.md
# README

Find newly created contests results at https://data.ijf.org.

## Require

* macOS or Linux.

Maybe Windows are also ok if babashka and diff are properly installed.
I'm using macOS to develop this script.

* babashka
https://github.com/babashka/babashka

* diff command.
The diff command is usually found at `/usr/bin/diff` in macos and Linux.

## Install
`compare_ijf.clj` を PATH の通ったフォルダに置き、実行可能ビットを立てておく。
名前は変えても大丈夫。以下は .clj を剥ぎ取った名前にセーブしてます。

    $ cp compare_ijf.clj ~/bin/compare_ijf
    $ chmod +x ~/bin/compare_ijf

## Usage

Output of compare_ijf is just the diff's output.
Each line shows the `id_completition`, `name` and `has_results` values
which are newly added on
https://data.ijf.org.

```
$ compare_ijf
> 1068,World Cup Buenos Aires 2012,100
> 1104,Panamerican Championships Seniors 2013,108
> 1106,African Championships Seniors 2013,97
> 1184,Oceanian Open Wollongong 2014,85
> 1185,Grand Prix Qingdao 2014,109
> 1186,Grand Prix Jeju 2014,112
```

For example, the first line tells that IJF updates the 100 results about the
competition whose id_competition is 1068, and whose name is 'World Cup Buenos
Aires 2012'.

## TODO

see CHANGELOG.md

## See also,

Followings are cousin projects of this. They are also under construction.

* fights https://github.com/hkim0331/fights
* fetch-ijf https://github.com/hkim0331/fetch-ijf

## License

Copyright &copy; 2022 Hiroshi Kimura


CHANGELOG.md
# CHANGELOG

## Unreleased
* ./competitions を問答無用で上書きしているのはどうか？ --no-update オプションとかは？
* --example オプション。嘘の .competitions と比較する。
* --dry-run オプション。~/.competitions をアップデートしない。
* --version オプション。
* diff が見つかったら自動的にダウンロードしてくるとか。
* diff のノイズ削減。下の行番号 68a69 はなくても問題ない。
```
68a69
> 1068,World Cup Buenos Aires 2012,100
```
* 一般化。compare-json
* nil はなに？1a2 は消すはず。
```
% ./compare_ijf.clj
nil
1a2
> 1001,World Cup Belo Horizonte 2010,112
```

## 0.3.4 - 2022-09-09
* --last オプション。最後にアップデートした IJF コンテストファイルの日付。

## 0.3.3 - 2022-09-09
* compare_ijf --version でバージョン表示。
* compare_ijf --update で比較の後にキープするデータをアップデート。
  新しいデータがなかったときはアップデートしない。
* compare_ijf 比較するが、キープするデータはそのまま。


## 0.3.1 - 2022-09-05
### Changed
* file-a, file-b を orig, download に名称変更。
* --update オプションがあり、かつ、diff が見つかった時だけ、
  orig を download で置き換える。

## 0.3.2 - 2022-09-05
### Added
- bump-version.sh やっぱあったほうがいい。

## 0.3.1 - 2022-08-19
### Changed
* namespaced. これに伴い、compare-ijf を compare_ijf に名称変更。
* updated README.md
### Fixed
* save_as_text appends "\n" at the end of file. This fixed 0.2.0 issue.

## 0.2.0 - 2022-08-18
* this is not a bug
```
! 999 World Cup Budapest 2010 56
--- 676,679 ----
! 999 World Cup Budapest 2010 56
\ No newline at end of file
```
emacs/vi appends a newline at the EOF if the file does not terminate
with newline.
### Changed
* save-as-text はコンマ区切りがお好みか？
* prefer `--normal` rather than `--context`?


## 0.1.0 - 2022-08-18
* project started.
* can display diffs.


compare.clj
#!/usr/bin/env bb
(require
  '[babashka.curl :as curl]
  '[cheshire.core :as json]
  '[clojure.java.io :as io]
  '[clojure.java.shell :refer [sh]]
  '[clojure.string :as str])

;; keep last competitions in a file.
(def ^:private competitions (str (System/getenv "HOME") "/.competitions"))
(def ^:private file-a "/tmp/competitions-A")
(def ^:private file-b "/tmp/competitions-B")

(defn get-url
  [url & [opt]] (curl/get url opt))

(defn get-competitions
  "Fetch competitions from data.ijf.org, return it as clojure map."
  []
  (-> (get-url "https://data.ijf.org/api/get_json"
               {:query-params {"params[action]" "competition.get_list"}})
      :body
      (json/parse-string true)))

(comment
  (def ijf (get-competitions)))

(defn filter-competitions
  "filter map remaining :id_competition, :name, :has_results.
   returns a map."
  [coll]
  (->> coll
       (map #(select-keys % [:id_competition :name :has_results]))
       (sort-by :id_competition)))

;;(filter-competitions ijf)

(defn save-as-text
  "save coll's values in `competitions`, as,
   1000 World Cup Cairo 2010 112
   1001 World Cup Belo Horizonte 2010 112
   1002 World Cup Madrid 2010 56
   ..."
  [coll]
  (let [lines (atom [])]
    (doseq [s coll]
      (swap! lines conj (str/join " " (vals s))))
    (spit competitions (str/join "\n" @lines))))

;;(save-as-text (filter-competitions ijf))

(defn compare-ijf
  "ダウンロード済みの competitions と新たにダウンロードする competitions を
   diff で比較する。"
  []
  (when (.exists (io/file competitions))
    (sh "cp" competitions file-a))
  (-> (get-competitions)
      (filter-competitions)
      (save-as-text))
  (sh "cp" competitions file-b)
  (let [{:keys [exit out err]} (sh "diff" "-c" file-a file-b)]
    (println out)))

(compare-ijf)

compare_ijf.clj
#!/usr/bin/env bb
(ns compare-ijf
  (:require
   [babashka.curl :as curl]
   [cheshire.core :as json]
   ;; [clojure.java.io :as io]
   [clojure.java.shell :refer [sh]]
   [clojure.string :as str]))

(def ^:private version "0.3.2")

;; keep last competitions in a file.
(def ^:private orig (str (System/getenv "HOME") "/.competitions"))
(def ^:private download "/tmp/from-ijf")

(defn get-url
  [url & [opt]] (curl/get url opt))

(defn get-competitions
  "Fetch competitions JSON from data.ijf.org,
   return it's body as clojure map."
  []
  (-> (get-url "https://data.ijf.org/api/get_json"
               {:query-params {"params[action]" "competition.get_list"}})
      :body
      (json/parse-string true)))

(defn filter-competitions
  "filter map remaining :id_competition, :name, :has_results.
   returns a map."
  [coll]
  (->> coll
       (map #(select-keys % [:id_competition :name :has_results]))
       (sort-by :id_competition)))

;;(filter-competitions ijf)

(defn save-as-text
  "save coll's values only to dest, as,
   1000 World Cup Cairo 2010 112
   1001 World Cup Belo Horizonte 2010 112
   1002 World Cup Madrid 2010 56
   ..."
  [coll dest]
  (let [lines (atom [])]
    (doseq [s coll]
      (swap! lines conj (str/join "," (vals s))))
    (spit dest (str/join "\n" @lines))
    (spit dest "\n" :append true)))

(defn compare-ijf
  [update?]
  (-> (get-competitions)
      (filter-competitions)
      (save-as-text download))
  (let [{:keys [exit out _]} (sh "diff" "--normal" "--text" orig download)]
    (if (zero? exit)
      (println "no new data")
      (do
        (println out)
        (when update?
          (sh "cp" download orig)
          (println "updated"))))))

;; if both --update option given and any diff found,
;; `orig` is replaced with `download.
(defn -main
  []
  (compare-ijf (some (partial = "--update")  *command-line-args*)))

(-main)


compare-ijf.clj
#!/usr/bin/env bb
;;;
;;; https://github.com/hkim0331/compare-ijf.git
;;;
(ns compare-ijf
  (:require
   [babashka.curl :as curl]
   [babashka.fs :as fs]
   [cheshire.core :as json]
   ;; [clojure.java.io :as io]
   [clojure.java.shell :refer [sh]]
   [clojure.string :as str]))

(def ^:private version "0.3.4")

;; keep last competitions in a file.
(def ^:private orig (str (System/getenv "HOME") "/.competitions"))
(def ^:private download "/tmp/from-ijf")

(defn get-url
  [url & [opt]] (curl/get url opt))

(defn get-competitions
  "Fetch competitions JSON from data.ijf.org,
   return it's body as clojure map."
  []
  (-> (get-url "https://data.ijf.org/api/get_json"
               {:query-params {"params[action]" "competition.get_list"}})
      :body
      (json/parse-string true)))

(defn filter-competitions
  "filter map remaining :id_competition, :name, :has_results.
   returns a map."
  [coll]
  (->> coll
       (map #(select-keys % [:id_competition :name :has_results]))
       (sort-by :id_competition)))

(defn save-as-text
  "save coll's values only to dest, as,
   1000 World Cup Cairo 2010 112
   1001 World Cup Belo Horizonte 2010 112
   1002 World Cup Madrid 2010 56
   ..."
  [coll dest]
  (let [lines (atom [])]
    (doseq [s coll]
      (swap! lines conj (str/join "," (vals s))))
    (spit dest (str/join "\n" @lines))
    (spit dest "\n" :append true)))

(defn compare-ijf
  [update?]
  (-> (get-competitions)
      (filter-competitions)
      (save-as-text download))
  (let [{:keys [exit out _]} (sh "diff" "--normal" "--text" orig download)]
    (if (zero? exit)
      (println "no new data")
      (do
        (println out)
        (when update?
          (sh "cp" download orig)
          (println "updated"))))))

(defn has?
  [s]
  (some #(= s %) *command-line-args*))

(defn mtime
  [f]
  (-> (fs/last-modified-time f)
      str))

(defn -main
  []
  (cond
    (has? "--version") (println version)
    (has? "--last") (mtime orig)
    :else (compare-ijf (has? "--update"))))

(-main)
